zeebe/atomix/cluster/src/test/java/io/atomix/raft/roles/RaftLeaderFlushErrorTest.java

```diff
@@ -0,0 +1,109 @@
+ * one or more contributor license agreements. See the NOTICE file distributed
+ * with this work for additional information regarding copyright ownership.
+ * Licensed under the Camunda License 1.0. You may not use this file
+ * except in compliance with the Camunda License 1.0.
+ */
+package io.atomix.raft.roles;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+import io.atomix.raft.FaultyFlusherConfigurator;
+import io.atomix.raft.RaftRule;
+import java.time.Duration;
+import java.util.Collection;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.stream.Stream;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+import org.junit.runners.Parameterized.Parameter;
+import org.junit.runners.Parameterized.Parameters;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@RunWith(Parameterized.class)
+public class RaftLeaderFlushErrorTest {
+
+  private static final Logger LOG = LoggerFactory.getLogger(RaftLeaderFlushErrorTest.class);
+  private static final int MEMBERS = 3;
+
+  @Parameter public boolean withDataLoss;
+
+  @Rule
+  @Parameter(1)
+  public RaftRule raftRule; // =
+
+  @Parameter(2)
+  public AtomicBoolean isFaulty;
+
+  @Parameter(3)
+  public AtomicInteger flushFailedCount;
+
+  @Parameters(name = "{index}: withDataLoss = {0}")
+  public static Collection<Object[]> parameters() {
+    return Stream.of(true, false)
+        .map(
+            withDataLoss -> {
+              final var isFaulty = new AtomicBoolean(false);
+              final var flushFailedCount = new AtomicInteger(0);
+              return new Object[] {
+                withDataLoss,
+                RaftRule.withBootstrappedNodes(
+                    MEMBERS,
+                    new FaultyFlusherConfigurator(
+                        (MEMBERS - 1) / 2,
+                        isFaulty::get,
+                        flushFailedCount::incrementAndGet,
+                        true,
+                        withDataLoss)),
+                isFaulty,
+                flushFailedCount
+              };
+            })
+        .toList();
+  }
+
+  @Test
+  public void shouldTransitionToFollowerWhenLeaderFailsToFlush() throws Exception {
+    // given
+    final var leader = raftRule.getLeader().get();
+    // check that the leader has a faulty flusher
+    assertThat(leader.name()).isEqualTo("1");
+
+    // get last index
+    raftRule.appendEntry();
+
+    // flusher fails to flush
+    isFaulty.set(true);
+
+    LOG.info("Leader flusher set to faulty");
+    // when
+    final var commitListener = raftRule.appendEntryAsync();
+
+    // then
+    assertThatThrownBy(() -> commitListener.awaitCommit(Duration.ofSeconds(2)))
```

Comment:
:x: :thinking: It's not ideal to just wait out until something doesn't happen. But then again, it's always hard to test the absence of something.

Maybe we can rethink the approach here. Maybe we can instead assert one or more of the following:

Leader steps down. Why not await a new leader? Since it had quorum, and it truncates entries on becoming a follower, then the other node (who gave quorum) should be elected, right? :thinking: 

That would also allow us to test that the old leader became follower, as expected.

Reply:
Normally we should elect the node with the longest log. But I now remember that we also truncate the uncommitted entries on becoming candidate, so I guess the other follower will also do the same, and then we will default to the priority instead.